src/connection.ts

```diff
@@ -967,42 +966,151 @@ export class Connection extends EventEmitter {
+        clearTimeout(timeout)
+        return packet
+      } catch (err) {
+        return null
+      }
+    }))
+
+    // Check if we got F08 errors (which will help determine the path maximum packet amount)
+    const maximumPacketAmounts = testPacketAmounts.map((sourceAmount, index) => {
+      if (results[index] && (results[index] as IlpPacket.IlpRejection).code === 'F08') {
+        try {
+          const reader = Reader.from((results[index] as IlpPacket.IlpRejection).data)
+          const receivedAmount = reader.readUInt64BigNum()
+          const maximumAmount = reader.readUInt64BigNum()
+          const maximumPacketAmount = new BigNumber(sourceAmount)
+            .times(maximumAmount)
+            .dividedToIntegerBy(receivedAmount)
+          this.log.debug(`sending test packet of ${testPacketAmounts[index]} resulted in F08 error that told us maximum packet amount is ${maximumPacketAmount}`)
+          return maximumPacketAmount
+        } catch (err) {
+          return new BigNumber(Infinity)
+        }
+      }
+      return new BigNumber(Infinity)
+    })
+    this.maximumPacketAmount = BigNumber.minimum(...maximumPacketAmounts.concat(this.maximumPacketAmount))
+    this.testMaximumPacketAmount = this.maximumPacketAmount
+    if (this.maximumPacketAmount.isEqualTo(0)) {
+      this.log.error(`cannot send anything through this path. the maximum packet amount is 0`)
+      throw new Error('Cannot send. Path has a Maximum Packet Amount of 0')
+    }
+
+    // Figure out which test packet discovered the exchange rate with the most precision
+    const { maxDigits, exchangeRate } = results.reduce(({ maxDigits, exchangeRate }, result, index) => {
+      if (result && (result as Packet).prepareAmount) {
+        const prepareAmount = (result as Packet).prepareAmount
+        const exchangeRate = prepareAmount.dividedBy(testPacketAmounts[index])
+        this.log.debug(`sending test packet of ${testPacketAmounts[index]} delivered ${prepareAmount} (exchange rate: ${exchangeRate})`)
+        if (prepareAmount.precision(true) >= maxDigits) {
+          return {
+            maxDigits: prepareAmount.precision(true),
+            exchangeRate
+          }
+        }
+      }
+      return { maxDigits, exchangeRate }
+    }, { maxDigits: 0, exchangeRate: new BigNumber(0) })
+
+    // TODO make the level of precision configurable
+    if (maxDigits >= 3) {
+      this.log.debug(`determined exchange rate to be ${exchangeRate}`)
+      this.exchangeRate = exchangeRate
+      return
+    }
+
+    // Try sending another packet if we don't have enough precision but ran into an F08 error
+    if (this.maximumPacketAmount.isFinite() && testPacketAmounts.indexOf(this.maximumPacketAmount.toNumber()) === -1) {
+      this.log.debug(`trying another packet with the maximum packet amount: ${this.maximumPacketAmount}`)
+      const result = await this.sendTestPacket(this.maximumPacketAmount)
+      // TODO what if this gets an F08 error?
```

Comment:
I was wondering this myself, maybe this should be attempted 3 times with decreasing amounts before giving up in the case of F08 errors. 

Another thought would be for it to just kick off the determine exchange rate function, but change it so that the parameter passed in are the amounts to send. So the first call would send in `[1, 1000, 1000000, 1000000000]` and then the second call would be like maximumPacketAmount / 4 = minimum, and then you pass in [minimum, minimum*2, minimum*3, and maximumPacketAmount. 

This would just be done one time, its recursive but if limited to one attempt it would basically be sending 2 volleys of packets in the event of an F08 before giving up.